go/libraries/doltcore/sqle/setalgebra/union.go

```diff
@@ -0,0 +1,342 @@
+		}
+
+		if !inRange {
+			hashToVal[h] = v
+		}
+	}
+
+	return CompositeSet{FiniteSet{hashToVal}, cs.Intervals}, nil
+}
+
+// intervalUnion takes two Interval objects and compares them then returns their union
+func intervalUnion(in1, in2 Interval) (Set, error) {
+	intComparison, err := compareIntervals(in1, in2)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return intervalUnionWithComparison(in1, in2, intComparison)
+}
+
+// intervalUnionWithComparison takes two Interval objects and their comparison and returns a new interval that
+// represents all the points in both intervals where possible, and returns a CompositeInterval when the two intervals
+// are non-overlapping.
+func intervalUnionWithComparison(in1, in2 Interval, intComparison intervalComparison) (Set, error) {
+	var resIntervToReduce Interval
+	if intComparison == noOverlapLess {
+		if in1.End != nil && in2.Start != nil && (in1.End.Inclusive || in2.Start.Inclusive) && in1.End.Val.Equals(in2.Start.Val) {
+			// in the case where you have intervals X and Y defined as A < X < B and B <= Y < C the comparison of the
+			// end of X and the start of Y will be -1.  But X includes all the points less than B, Y includes B and the
+			// points up until C.  So the resulting interval Z would be A < Z < C.
+			resIntervToReduce = Interval{in1.nbf, in1.Start, in2.End}
+		} else {
+			// Non overlapping intervals. Create CompositeSet with intervals in sorted order.
+			return CompositeSet{FiniteSet{make(map[hash.Hash]types.Value)}, []Interval{in1, in2}}, nil
+		}
+	} else if intComparison == noOverlapGreater {
+		if in2.End != nil && in1.Start != nil && (in2.End.Inclusive || in1.Start.Inclusive) && in2.End.Val.Equals(in1.Start.Val) {
+			// see above for info no this case
+			resIntervToReduce = Interval{in1.nbf, in2.Start, in1.End}
+		} else {
+			// Non overlapping intervals. Create CompositeSet with intervals in sorted order.
+			return CompositeSet{FiniteSet{make(map[hash.Hash]types.Value)}, []Interval{in2, in1}}, nil
+		}
+	} else if intComparison[start1start2] <= 0 {
+		if intComparison[end1end2] >= 0 {
+			// the first interval wholly contains the second. Return the first.
+			return in1, nil
+		} else {
+			// return an interval with the smallest start point and largest end point
+			resIntervToReduce = Interval{in1.nbf, in1.Start, in2.End}
+		}
+	} else {
+		if intComparison[end1end2] <= 0 {
+			// the second interval wholly contains the first. Return the second.
+			return in2, nil
+		} else {
+			// return an interval with the smallest start point and largest end point
+			resIntervToReduce = Interval{in1.nbf, in2.Start, in1.End}
+		}
+	}
+
+	return simplifyInterval(resIntervToReduce)
+}
+
+// intervalCompositeSetUnion will check the CompositeSet's FiniteSet for points that the new Interval contains
+// and exclude those from the resulting composite and then union the Interval with its existing intervals.
+func intervalCompositeSetUnion(in Interval, cs CompositeSet) (Set, error) {
```

Comment:
Can this be Union(cs.fs, UnionWithMultiples(in,  cs.in...)

Reply:
same as finiteSetCompositeSetUnion